internal/confmapprovider/discovery/discovery.go

```diff
@@ -0,0 +1,123 @@
+// Copyright Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/collector/confmap"
+	"go.uber.org/zap"
+
+	"github.com/signalfx/splunk-otel-collector/internal/settings"
+)
+
+var _ confmap.Provider = (*providerShim)(nil)
+
+type Provider interface {
+	ConfigDScheme() string
+	ConfigDProvider() confmap.Provider
+	DiscoveryModeScheme() string
+	DiscoveryModeProvider() confmap.Provider
+}
+
+type providerShim struct {
+	retrieve func(ctx context.Context, uri string, watcher confmap.WatcherFunc) (*confmap.Retrieved, error)
+	scheme   string
+}
+
+func (p providerShim) Retrieve(ctx context.Context, uri string, watcher confmap.WatcherFunc) (*confmap.Retrieved, error) {
+	return p.retrieve(ctx, uri, watcher)
+}
+
+func (p providerShim) Scheme() string {
+	return p.scheme
+}
+
+func (p providerShim) Shutdown(ctx context.Context) error {
+	return nil // nolint:unparam
+}
+
+type mapProvider struct {
+	logger  *zap.Logger
+	configs map[string]*Config
+}
+
+func New() (Provider, error) {
+	m := &mapProvider{configs: map[string]*Config{}}
+	zapConfig := zap.NewProductionConfig()
+	zapConfig.Level = zap.NewAtomicLevelAt(zap.DebugLevel)
+	var err error
+	if m.logger, err = zapConfig.Build(); err != nil {
+		return (*mapProvider)(nil), err
```

Comment:
so this will only return a fatal error if the zap error is not nil, how likely is that? Could we swallow the error instead? The log.Fatalf earlier seemed scary.

Reply:
Errors will occur if the default encoder fails to build, which would be very unlikely assuming there's not a larger zap issue that would likely be uncovered by all the integration tests. This logger and its core are probably going to be used in discovery mode depending on the state of my refactoring, so it's a pretty important though unlikely situation.